{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bfcebe0d-7ce1-4e9e-9c77-9b575eaa5194",
   "metadata": {},
   "source": [
    "## Initializing the experiment\n",
    "\n",
    "First, create the 'common' part of the dyngen simulations as follows. This will ensure that the gene regulatory network part of the simulation is exactly the same."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2063c316-0ac8-4ca6-8c8f-55b1799278e9",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: dynfeature\n",
      "\n",
      "Loading required package: dynguidelines\n",
      "\n",
      "Loading required package: dynmethods\n",
      "\n",
      "Loading required package: dynplot\n",
      "\n",
      "Loading required package: dynwrap\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "slurmR default option for `tmp_path` (used to store auxiliar files) set to:\n",
      "  /mnt/scratcha/fmlab/baker02/tidyscreen/notebooks\n",
      "You can change this and checkout other slurmR options using: ?opts_slurmR, or you could just type \"opts_slurmR\" on the terminal.\n",
      "\n",
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.1 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.3     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.1     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.6\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.1.3     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 1.4.0     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n",
      "\n",
      "Attaching package: ‘assertthat’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tibble’:\n",
      "\n",
      "    has_name\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    transpose\n",
      "\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(dyno)\n",
    "library(slurmR)\n",
    "library(dyngen)\n",
    "library(parallel)\n",
    "library(dynutils)\n",
    "library(tidyverse)\n",
    "library(assertthat)\n",
    "library(data.table)\n",
    "library(SingleCellExperiment)\n",
    "source(\"../scripts/perturbation.R\")\n",
    "source(\"../scripts/backbone_cellpop.R\")\n",
    "\n",
    "set.seed(1)\n",
    "data(\"realcounts\", package = \"dyngen\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "578743c5-3500-4f6e-8a85-bc4c54d97bce",
   "metadata": {},
   "outputs": [],
   "source": [
    "data(\"realcounts\", package = \"dyngen\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8283d0a5-f114-4495-82df-80f270f4279d",
   "metadata": {},
   "outputs": [],
   "source": [
    "backbone <- backbone_bifurcating()\n",
    "\n",
    "config = initialise_model(\n",
    "            backbone = backbone,\n",
    "            num_cells = 1000,\n",
    "            num_targets = 500,\n",
    "            num_hks = 500,\n",
    "            verbose = interactive(),\n",
    "            download_cache_dir = tools::R_user_dir(\"dyngen\", \"data\"),\n",
    "            simulation_params = simulation_default(\n",
    "              total_time = 1000,\n",
    "              census_interval = 2, \n",
    "              ssa_algorithm = ssa_etl(tau = 300/3600),\n",
    "              experiment_params = simulation_type_wild_type(num_simulations = 10)\n",
    "            ),\n",
    "            experiment_params = experiment_snapshot(\n",
    "              realcount = \"GSE100866_CBMC_8K_13AB_10X-RNA_umi\"\n",
    "            )\n",
    "          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "933e5017-873e-412e-8b0a-80aee7fa9ae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# backbone <- backbone_bifurcating()\n",
    "\n",
    "# config <-\n",
    "#   initialise_model(\n",
    "#     backbone = backbone,\n",
    "#     num_cells = 1000,\n",
    "#     num_tfs = nrow(backbone$module_info),\n",
    "#     num_targets = 250,\n",
    "#     num_hks = 250,\n",
    "#     simulation_params = simulation_default(\n",
    "#       census_interval = 10, \n",
    "#       ssa_algorithm = ssa_etl(tau = 300 / 3600),\n",
    "#       experiment_params = simulation_type_wild_type(num_simulations = 100)\n",
    "#     )\n",
    "#   )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eeda2849-504d-4e93-9a1c-9287c46c5253",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "model_common = config %>%\n",
    "                  generate_tf_network() %>%\n",
    "                  generate_feature_network() %>% \n",
    "                  generate_kinetics() %>%\n",
    "                  generate_gold_standard()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4101279c-47f7-4e86-a8ec-ede5d62df172",
   "metadata": {},
   "source": [
    "## Simulate Wild Type\n",
    "\n",
    "Simulating the wild type is to simply run the standard dyngen way of simulating cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7da62e63-f8a7-40d6-aced-ed771a455607",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_wt = model_common %>% generate_cells()\n",
    "plot_gold_mappings(model_wt, do_facet = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "533163dd-c4d8-475c-ba88-1a84242fa61e",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_backbone_modulenet(model_common)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80191081-9475-4e89-a6c3-f07675726870",
   "metadata": {},
   "source": [
    "## Simulate Knockout\n",
    "\n",
    "To simulate a gene knockout, the experiment type of the simulation is changed from the wild type to knockdown. The **simulation_type_knockdown** function is actually designed to perform single-cell knockdowns of 1 to 5 random genes by default! But this behavior can be overriden by setting the following parameters:\n",
    "\n",
    "- **genes**: list of gene ids to be knocked out\n",
    "- **num_genes**: number of genes to be knocked out per gene (sampled randomly)\n",
    "- **multiplier**: the fraction of the transcription rate after KO (0 = no transcription, 1 = no Knockout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3949bc32-7dbe-41ac-a608-6277de86ddf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "b3_genes = model_common$feature_info %>% filter(module_id == \"B3\") %>% pull(feature_id)\n",
    "model_ko = model_common\n",
    "\n",
    "model_ko$simulation_params$experiment_params = simulation_type_knockdown(\n",
    "  num_simulations = 100L,\n",
    "  timepoint = 0, \n",
    "  genes = b3_genes,\n",
    "  num_genes = length(b3_genes),\n",
    "  multiplier = 0\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2a5890a-bf3b-4b71-ad4d-77c1ec69faf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_ko = model_ko %>% generate_cells()\n",
    "plot_gold_mappings(model_ko, do_facet = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e981e5e-c399-42fb-95a7-c34018d52ca0",
   "metadata": {},
   "source": [
    "## Combine Outputs and Visualise\n",
    "\n",
    "We can combine both simulations into one simulation and visualize the output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f51e9ef-cd9b-477b-9871-4dadf898ecfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_comb = combine_models(list(WT = model_wt, KO1 = model_ko)) %>% generate_experiment()\n",
    "plot_simulations(model_comb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dae9a4f8-4a5c-430d-ba83-18e05a689458",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_gold_mappings(model_comb, do_facet = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e743a387-a2a1-4264-b9ca-69010037860b",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment = model_comb$experiment"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
